provider/aws/aws_test.go

```diff
@@ -807,22 +827,101 @@ func TestAWSsubmitChangesError(t *testing.T) {
+	ep2 := endpoint.NewEndpointWithTTL("fail.zone-1.ext-dns-test-2.teapot.zalan.do", endpoint.RecordTypeA, endpoint.TTL(recordTTL), "1.0.0.2")
+	ep3 := endpoint.NewEndpointWithTTL("success2.zone-1.ext-dns-test-2.teapot.zalan.do", endpoint.RecordTypeA, endpoint.TTL(recordTTL), "1.0.0.3")
+
+	ep2txt := endpoint.NewEndpointWithTTL("fail__edns_housekeeping.zone-1.ext-dns-test-2.teapot.zalan.do", endpoint.RecordTypeTXT, endpoint.TTL(recordTTL), "something") // "__edns_housekeeping" is the TXT suffix
+	ep2txt.Labels = map[string]string{
+		endpoint.OwnedRecordLabelKey: "fail.zone-1.ext-dns-test-2.teapot.zalan.do",
+	}
+
+	// "success" and "fail" are created in the first step, both are submitted in the same batch; this should fail
+	cs1 := provider.newChanges(route53.ChangeActionCreate, []*endpoint.Endpoint{ep2, ep2txt, ep1})
+	input1 := &route53.ChangeResourceRecordSetsInput{
+		HostedZoneId: aws.String("/hostedzone/zone-1.ext-dns-test-2.teapot.zalan.do."),
+		ChangeBatch: &route53.ChangeBatch{
+			Changes: cs1.Route53Changes(),
+		},
+	}
+	clientStub.MockMethod("ChangeResourceRecordSets", input1).Return(nil, fmt.Errorf("Mock route53 failure"))
+
+	// because of the failure, changes will be retried one by one; make "fail" submitted in its own batch fail as well
+	cs2 := provider.newChanges(route53.ChangeActionCreate, []*endpoint.Endpoint{ep2, ep2txt})
+	input2 := &route53.ChangeResourceRecordSetsInput{
+		HostedZoneId: aws.String("/hostedzone/zone-1.ext-dns-test-2.teapot.zalan.do."),
+		ChangeBatch: &route53.ChangeBatch{
+			Changes: cs2.Route53Changes(),
+		},
+	}
+	clientStub.MockMethod("ChangeResourceRecordSets", input2).Return(nil, fmt.Errorf("Mock route53 failure"))
+
+	// "success" should have been created, verify that we still get an error because "fail" failed
+	require.Error(t, provider.submitChanges(ctx, cs1, zones))
+
+	// assert that "success" was successfully created and "fail" and its TXT record were not
+	records, err := provider.Records(ctx)
+	require.NoError(t, err)
+	ok := false
+	for _, r := range records {
+		switch r.DNSName {
+		case "success.zone-1.ext-dns-test-2.teapot.zalan.do":
+			ok = true
+		case "fail.zone-1.ext-dns-test-2.teapot.zalan.do", "fail__edns_housekeeping.zone-1.ext-dns-test-2.teapot.zalan.do":
+			ok = false
+			break
+		}
+	}
+	require.True(t, ok)
+
+	// next batch should contain "fail" and "success2", should succeed this time
+	cs3 := provider.newChanges(route53.ChangeActionCreate, []*endpoint.Endpoint{ep2, ep2txt, ep3})
+	require.NoError(t, provider.submitChanges(ctx, cs3, zones))
+
+	// verify all records are there
+	records, err = provider.Records(ctx)
+	require.NoError(t, err)
+	count := 0
+	for _, r := range records {
+		switch r.DNSName {
+		case
+			"success.zone-1.ext-dns-test-2.teapot.zalan.do",
+			"fail.zone-1.ext-dns-test-2.teapot.zalan.do",
+			"success2.zone-1.ext-dns-test-2.teapot.zalan.do",
+			"fail__edns_housekeeping.zone-1.ext-dns-test-2.teapot.zalan.do":
+			count += 1
```

Comment:
I think it would be better to have a function that checks whether a particular record, given its name, is part of a list of records.

This way we could check the presence (and absence for above) more explicitly and avoid the obscure `count` and `ok` variables, e.g.

```go
// check inclusion
require.True(t, containsRecordWithDNSName("success.zone-1.ext-dns-test-2.teapot.zalan.do", records)
// check exclusion
require.False(t, containsRecordWithDNSName("fail.zone-1.ext-dns-test-2.teapot.zalan.do", records)
```